### ISSUE TITLE
Support for responsive font_size with relative length units - (em % vw vim vmax etc.)

### ISSUE THREAD

**Issue by ZenithStar** [godotengine/godot-proposals#9746]
### Describe the project you are working on

UI-heavy Mobile Game

### Describe the problem or limitation you are having in your project

One of the basic features of a responsive UI is that the font size rasters dynamically proportional to the viewport size. I've see many people recommend solving this by using scaling, leading to many more users searching for ways to make their scaled fonts appear sharper.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

HTML+CSS solves this by offering several relative [unit lengths])(https://developer.mozilla.org/en-US/docs/Web/CSS/length) for the font-size, whereas Godot only supports pixel length.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I'm not sure what the most minimally intrusive implementation would be, but I see two options.

- Change the type of any integer font_size field of Theme and any Control Theme Property overrides to be of type string. This string is evaluated by Theme into an integer px length. Control nodes implementing a font_size property will also need to connect to Viewport.size_changed to trigger a reevaluation.

- Change the type of any integer font_size field of Theme and any Control Theme Property overrides to be of a new FontSize Resource type. This resource holds a string expression and evaluates the expression into an integer px length. The FontSize resource also connects itself to Viewport.size_changed to trigger a reevaluation.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can be used often. It is used in virtually every modern GUI application, including the one you're reading this message on right now.

### Is there a reason why this should be core and not an add-on in the asset library?

This is core. I can conceive of a workaround addon that supports this feature, but it would involve a server that maintains a list of references to every font_size field in the scene tree.

**Comment by Calinou** [comment]
> One of the basic features of a responsive UI is that the font size rasters dynamically proportional to the viewport size. I've see many people recommend solving this by using scaling, leading to many more users searching for ways to make their scaled fonts appear sharper.

Godot already offers a built-in solution for this that doesn't involve scaling Control nodes: the `canvas_items` stretch mode. See [Multiple resolutions](https://docs.godotengine.org/en/stable/tutorials/rendering/multiple_resolutions.html) in the documentation :slightly_smiling_face:

**Comment by ZenithStar** [comment]
> If you look at most how modern AAA games scale their UI according to resolution, it's how they do it.

Most AAA titles offer support for only a very small number of resolutions and aspect ratios, which works, but isn't necessarily ideal. Most of them are designed primarily with console and controller in mind. Most PC-focused games will not scale the UI by default, but instead offer a slider bar in the options menu for the player to adjust the scaling. Some games that do this off the top of my head are Path of Exile, Terraria, and Age of Empires II:DE.

Mobile apps typically do not have this luxury, as they cater to a less-tech-savvy audience, while simultaneously needing to deal with more extreme scaling requirements (e.g. iPhone 15 Pro is 393x852 while iPad Pro 12.9" is 1024x1366). This is especially true for an app that supports auto-rotating between portrait and landscape layouts. Godot advertises its ability to create responsive UIs, but is missing certain features like this one that I would consider core to a responsive toolkit.

**Comment by Calinou** [comment]
> Most PC-focused games will not scale the UI by default, but instead offer a slider bar in the options menu for the player to adjust the scaling.

You can do that while using the `canvas_items` stretch mode using [stretch scale](https://docs.godotengine.org/en/stable/tutorials/rendering/multiple_resolutions.html#stretch-scale).

Please test the [Multiple Resolutions and Aspect Ratios](https://github.com/godotengine/godot-demo-projects/tree/master/gui/multiple_resolutions) demo, it covers stuff like that.

**Comment by ZenithStar** [comment]
I'm well aware of the various solutions involving scaling, but it's ultimately a hack that leads to either artifacting and/or unnecessary overdrawing. Arguably, the most complicated function in this pipeline is the rasterization function, which is fully implemented for any pixel font_size value, but the upstream from that function is underutilizing its potential by only passing static font sizes.

**Comment by Meorge** [comment]
Ah, I wrote a proposal for this and only thought after my draft was complete to see if something already existed 😅 

Here is a snippet from my proposal regarding a (partial) implementation of this, specifically for RichTextLabel at least:

> Add the `em` (and possibly `rem`) units for the `font_size` tag.
> 
> The `em` unit would set the enclosed text's font size relative to its parent text.
> With a RichTextLabel with a default font size of 50px, the following block:
> ```
> [font_size=30px]Hello [font_size=0.5em]World[/font_size][/font_size] how are you?
> ```
> would display "Hello" at 30px, "World" at 15px, and "how are you" at 50px.
> 
> The `rem` unit would set the enclosed text's font size relative to the RichTextLabel's default font size.
> With a RichTextLabel with a default font size of 50px, the following block:
> ```
> [font_size=30px]Hello [font_size=0.5rem]World[/font_size][/font_size] how are you?
> ```
> would display "Hello" at 30px, "World" at 25px, and "how are you" at 50px.

> [!note]
> You might be asking, **"Why not use the percent symbol like `[font_size=50%]` instead of `[font_size=0.5em]`?"**
>
> That's a good question. I was even motivated to file this proposal myself when I tried using percents and it didn't work! The reason I've proposed `em` here is that **the `[hr]` and `[img]` tags already use the percent symbol to specify percentages of *the control width***. For consistency, I think it would be best to use the `em` unit as relative to the current font size following https://github.com/godotengine/godot/pull/112617 .

**Comment by rempas** [comment]
Yes, yes, yes! I want to use Godot to make Desktop/Mobile apps and this is needed so much! I know that Godot is not the most *ideal* tool for creating stand-alone apps because it's a bit bloated for that matter but, it's still much better than Electron and other web-based crap.
